daal4py/sklearn/neighbors/_base.py

```diff
@@ -0,0 +1,278 @@
+    estimator.n_samples_fit_ = X.shape[0]
+    estimator.n_features_in_ = X.shape[1]
+    estimator._fit_X = X
+    estimator._fit_method = estimator.algorithm
+    estimator.effective_metric_ = 'euclidean'
+
+    params = {
+        'method': 'defaultDense',
+        'k': estimator.n_neighbors,
+        'nClasses': len(estimator.classes_),
+        'voteWeights': 'voteUniform' if estimator.weights == 'uniform' else 'voteDistance',
+        'resultsToEvaluate': 'computeClassLabels',
+        'resultsToCompute': ''
+    }
+
+    method = parse_auto_method(estimator, estimator.algorithm, estimator.n_samples_fit_, estimator.n_features_in_)
+    train_alg = training_algorithm(method, fptype, params)
+    estimator.daal_model_ = train_alg.compute(X, estimator._y.reshape(-1, 1)).model
+
+
+def daal4py_kneighbors(estimator, X=None, n_neighbors=None, return_distance=True):
+    n_features = getattr(estimator, 'n_features_in_', None)
+    shape = getattr(X, 'shape', None)
+    if n_features and shape and len(shape) > 1 and shape[1] != n_features:
+        raise ValueError('Input data shape {} is inconsistent with the trained model'.format(X.shape))
+
+    check_is_fitted(estimator)
+
+    if n_neighbors is None:
+        n_neighbors = estimator.n_neighbors
+    elif n_neighbors <= 0:
+        raise ValueError(
+            "Expected n_neighbors > 0. Got %d" %
+            n_neighbors
+        )
+    else:
+        if not isinstance(n_neighbors, numbers.Integral):
+            raise TypeError(
+                "n_neighbors does not take %s value, "
+                "enter integer value" %
+                type(n_neighbors))
+
+    if X is not None:
+        query_is_train = False
+        X = check_array(X, accept_sparse='csr')
+    else:
+        query_is_train = True
+        X = estimator._fit_X
+        # Include an extra neighbor to account for the sample itself being
+        # returned, which is removed later
+        n_neighbors += 1
+
+    n_samples_fit = estimator.n_samples_fit_
+    if n_neighbors > n_samples_fit:
+        raise ValueError(
+            "Expected n_neighbors <= n_samples, "
+            " but n_samples = %d, n_neighbors = %d" %
+            (n_samples_fit, n_neighbors)
+        )
+
+    chunked_results = None
+
+    try:
+        fptype = getFPType(X)
+    except ValueError:
+        fptype = None
+
+    params = {
+        'method': 'defaultDense',
+        'k': n_neighbors,
+        'resultsToCompute': 'computeIndicesOfNeighbors',
+        'resultsToEvaluate': 'none'
+    }
+    if return_distance:
+        params['resultsToCompute'] += '|computeDistances'
+
+    method = parse_auto_method(estimator, estimator._fit_method, estimator.n_samples_fit_, n_features)
+
+    predict_alg = prediction_algorithm(method, fptype, params)
+    prediction_result = predict_alg.compute(X, estimator.daal_model_)
+
+    if return_distance:
+        results = prediction_result.distances.astype(fptype), prediction_result.indices.astype(int)
```

Comment:
For some reason, kd_tree always returns distances as float64, brute - depends on X from fit().